src/pass/verify_memory.cc

```diff
@@ -0,0 +1,166 @@
+/*!
+ *  Copyright (c) 2018 by Contributors
+ * \file verify_memory.cc
+ * \brief Pass to check if memory accesses are legal.
+ */
+#include <tvm/ir.h>
+#include <tvm/ir_visitor.h>
+#include <tvm/ir_pass.h>
+#include <tvm/operation.h>
+
+namespace tvm {
+namespace ir {
+namespace {
+
+/*!
+ * \brief Verify if memory accesses are legal.
+ *
+ *  In the case that tgt is cuda, if workload is not bound with
+ *  threads, CPU code is generated that tries to access GPU memory,
+ *  which is illegal.
+ *
+ *  This pass performs such verification by checking if all Producer/Consumer
+ *  with memory accesses are bound with threads when device type is GPU.
+ */
+class MemoryAccessVerifier final : protected IRVisitor {
+ public:
+  /// Special member functions
+  //@{
+  explicit MemoryAccessVerifier(LoweredFunc f, DLDeviceType device_type)
+      : func_(f), dev_type_(device_type) {}
+  virtual ~MemoryAccessVerifier() = default;
+  MemoryAccessVerifier(const MemoryAccessVerifier &) = delete;
+  MemoryAccessVerifier(MemoryAccessVerifier &&) = delete;
+  MemoryAccessVerifier &operator=(const MemoryAccessVerifier &) = delete;
+  MemoryAccessVerifier &operator=(MemoryAccessVerifier &&) = delete;
+  //@}
+
+  /// Interface to perform memory access verification
+  void Run() {
+    if (kDLGPU != dev_type_) return;
```

Comment:
We need this for most GPU devices. Including OpenCL, Vulkan, Metal,

Reply:
Let us add a IsGPUDevice function which returns true it it is GPU and use it in here

Reply:
Hi Tianqi,

Thank you for the comments. I am new to TVM, may I know what DLDeviceType values denote GPU devices, please?

For all possible values, what I get from dlpack.h are: 
  kDLCPU = 1,
  kDLGPU = 2,
  kDLCPUPinned = 3,
  kDLOpenCL = 4,
  kDLMetal = 8,
  kDLVPI = 9,
  kDLROCM = 10,

Which one is for the Vulkan case?

Reply:
Vulkan is not yet upstreamed. https://github.com/dmlc/tvm/blob/d5c42b0c29ac4793c9cd0f09d8a733981b56eb50/include/tvm/runtime/c_runtime_api.h#L59

Reply:
So we need Metal, OpenCL, ROCM, Vulkan and kDLGPU, kOpenGL

Reply:
I have updated the code according to your comments. Many thanks to this comprehensive list. It saved my day.